src/formatter.ts

```diff
@@ -0,0 +1,78 @@
+import * as cp from "child_process";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { TextEncoder } from "util";
+
+export class Formatter implements vscode.DocumentFormattingEditProvider {
+    readonly binaryPath: string = '';
+
+    constructor(binaryPath: string) {
+        if (!binaryPath || binaryPath.length === 0) {
+            throw new Error('binaryPath is required to construct a formatter');
+        }
+        this.binaryPath = binaryPath;
+    }
+
+    public provideDocumentFormattingEdits(
+        document: vscode.TextDocument,
+        // @ts-ignore: The formatter doesn't have any options, but we need to implement the interface.
+        options: vscode.FormattingOptions,
+        token: vscode.CancellationToken
+    ): vscode.ProviderResult<vscode.TextEdit[]> {
+        if (vscode.window.visibleTextEditors.every((e) => e.document.fileName !== document.fileName)) {
+            return [];
+        }
+        return this.runFormatter(document, token).then(
+            (edits) => edits,
+            (err) => {
+                console.log(err);
+                return Promise.reject('Check the console in dev tools to find errors when formatting.');
+            }
+        );
+    }
+
+    private randomId(): string {
+        return (Math.floor(Math.random() * 100000)).toString()
```

Comment:
Don't forget to fix lint here (the missing `;`).

Reply:
Yup, the multiplication is just so that we get a whole instead of a float (that and the `Math.floor`). the `floor` guarantees this will be a whole number :)

I at first tried getting some sort of uuid, and looked at some diff libraries but they all had diff issues
* [unique-file-name](https://www.npmjs.com/package/unique-file-name) and [unique-filename](https://www.npmjs.com/package/unique-filename) don't have ts support
* [uniquefilename](https://www.npmjs.com/package/uniquefilename) only does sequential numbers
* [nanoid](https://www.npmjs.com/package/nanoid) just doesn't work? I get some errors when loading the extension "failed to load nanoid" despite it being stalled both globally and locally :(

I figured `100...0` was enough zeroes, so it wasn't worth it spending anymore time learning js. But I added a couple more zeros just in case.